### PR TITLE
Implementing ViT for ADNI Dataset Classification - 47028643

### PR DESCRIPTION
**Pull request by Fhall21**
### Description
For this pull request, I am submitting my solution to the challenge of implementing a Visual Transformer (ViT) model to classify ADNI dataset. In the end I achieved an accuracy of 66.94% on the test dataset.

### Files and Folders Overview

Main files:
- **modules.py**: This file contains the code for the ViT model architecture and implementation.
- **dataset.py**: Code for load the dataset, as well as calculate the mean and standard deviation of the dataset.
- **train.py**: The primary script for training the model on the ADNI dataset, it is used to train, validate, test and save the model.
- **predict.py**: Script for evaluating the model's performance on and producing visuals from the test data.
- **driver.py**: This file is used to run the training and prediction scripts.

Relating to the README:
- **README.md**: A comprehensive README.md providing information about the model, the training process, and results.
- **/train_val_loss_curves/**: This folder contains the train and validation loss curves from hyperparameter tuning which is refered to within a table within the README.md
- **/example_train_images/**: This folder contains an example of the images used for training the model - also referred to within the README.md
- **/saved_models/train_visuals/45**: This folder path was the folder path used when saving model outputs, in this case the output visualisations are saved and the model file itself has been removed.
- **training_and_validation_loss_45.png**: This file contains the training and validation loss curves from the final training run of the model.
- **ViT-architecture.png**: This file contains an image of the ViT architecture used for this project.

### Note on Final Results
Overall, despite large hyperparameter tuning the model was not able to get much higher than 66%. In future, other variants of the ViT may be considered so as to gain such, but for the given architecture and hyperparameters, as evident by the testing table in the README.md, this was the best result I was able to achieve. Many of the other models tried tended to overfit and this was the best model when considering such as well as computational time (which realistically is a key resource but was not a key part of this challenge).

### Author

Felix Hall
47028643

Thank you for considering this pull request.

### CODE REVIEW

**Comment by aryaman-sh** [comment]
**This is an initial inspection, no action is required at this point**

**Difficulty**
- Hard
- Task 6

**Readme**
- Excellent

**Commit messages**
- good

**Code**
- Uses vision transformer
- Good structure
- Good comments

**Functionality/Performance**
- accuracy 66.8%

**General comments**
- Good hyperparameter tuning
- No changes required

**Comment by shakes76** [comment]
# Marking
## Good Practice (Design/Commenting, TF/Torch Usage)
Adequate design and implementation
Good spacing and comments
Header blocks
## Recognition Problem 
Solves problem 
Driver Script present
File structure present
Shows Usage & Demo & Visualisation & Data usage
Module present
Commenting
No Data leakage
Difficulty: Hard
## Commit Log
Meaningful commit messages
Progressive commits used
## Documentation
ReadMe good
Model/technical explanation
Good Description and Comments 
Markdown used and PDF submitted
## Pull Request
Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch) 
No Feedback required
Request Description good

**Comment by shakes76** [comment]
Cannot merge because repo README deleted. Please remove DSTORE files as well. Does not affect grade only the merge.

**Comment by wangzhaomxy** [comment]
No feedback attempt and no feedback marks lost.

**Comment by wangzhaomxy** [comment]
No feedback attempt and no feedback marks lost.

**Comment by Fhall21** [comment]
I have added the repo README (root folder README file which I assumed was the one you were referring to), as well as deleted the DSTORE files from each folder.

**Comment by wangzhaomxy** [comment]
Feedback attempt and no feedback marks lost.
